### PR TITLE
Disable CSE for the initialization function

### PR DESCRIPTION
**Pull request by chambart**
This is #1453 with only the CSE change. This is sufficient for compiling the reported problematic files of https://caml.inria.fr/mantis/view.php?id=7630 with flambda with a few hundreds MB of ram while it took ~10GB without this patch.

### CODE REVIEW

File: asmcomp/asmgen.ml
Comment (gasche) [inline asmcomp/asmgen.ml]: You should link at a relevant MPR to justify the choice of disabling specifically CSE.

File: asmcomp/cmmgen.ml
Comment (gasche) [inline asmcomp/cmmgen.ml]: You should link to the relevant MPR to justify the specific choice of enabling `fast_compile` for compilation-unit functions.

**Comment by xavierleroy** [comment]
I'm unsure whether we need the new field `fun_fast_compile`: wouldn't the negation of the existing field `fun_fast` work for this purpose?

**Comment by chambart** [comment]
It's not exactly the same thing, for instance, I know that the frama-c guys set it to false to reduce code size.
Disabling CSE in that case would be counterproductive.

**Comment by alainfrisch** [comment]
- I'm unsure about the name of the new field.  At some point, we might want to add precise control over compilation options (disabling other passes, choice of register allocator, etc) at the function level (for instance with annotation, or other heuristics), and `fun_fast_compile` would be too coarse grained.  I suggest `fun_cse_disabled` (or fun_cse_enabled).

- Is the problem really specific to flambda?  I.e. does the "transl_store" compilation strategy for non-flambda avoid the problem?  If so, what about disabling cse for the entry function only in flambda mode?

- From https://caml.inria.fr/mantis/view.php?id=7630, I understand that the problem comes from sharing loads of addresses of global symbols.  Cannot this happen in arbitrary code?

**Comment by chambart** [comment]
* I have no objection to changing the name. Note that a curated set of options that favor compilation time
  against performance per function might be useful. For instance, if linscan is deemed stable enought at
  some point in the future, it would fit such a criteria.

* The real problem triggers only with flambda, but the compilation time benefit is noticeable with clambda
  too while not degrading performances (unless you wrote your computation heavy loop at toplevel).
  The main difference between flambda an non-flambda in that case is the number of symbols. The symbol
  address loading is shared too, but as there is only one, the interaction is not too problematic.

* I never encountered a function remotely as large as what can appear at toplevel. Something similar
  could appear if you turned your large compilation unit into a functor for instance, but in that situation CSE
  wouldn't be the main problem.

Such an option could be available through an annotation per function. I wouldn't like something too low level that explicitly disable passes. Such an option would imply `[@inline never]` as inlining would erase the annotation.

**Comment by chambart** [comment]
This PR is suggested for 4.06.1. For the bugfix release, this would only apply to flambda to minimize the impact.

This allows uunf to build (see http://obi.ocamllabs.io/logs/eedf08d88b07040656ce206266ba0c82.txt) with flambda.

**Comment by xavierleroy** [comment]
I'm OK with the basic approach but like @alainfrisch I have doubts about the name of the new field, and moreover I have doubts that we should add a new Boolean field for every variation in code generation we need.  

What about a single field `fun_codegen_options` of type `codegen_option list`?  The type `codegen_option` would contain constructors such as `CG_no_CSE`.  The existing `fun_fast` field could also be removed and turned into a `codegen_option`, perhaps by reversing the logic (no flag = optimise for speed; `CG_reduce_code_size` = optimise for size).  The general idea is that the absence of flags would correspond to the most common code generation choices, e.g. "optimise for speed", "perform CSE", etc.

**Comment by xavierleroy** [comment]
Also the issue may not be hot enough to go in a possible 4.06.1 release.  At any rate let's shoot for 4.07.

**Comment by ejgallego** [comment]
Indeed with this patch the Coq problems seem solved, thanks to all!

Informal timings:
```
A = 4.06.1+flambda+pr1455 "-O3 -unbox-closures"
B = 4.06.1+flambda+pr1455 "-O3 -unbox-closures -native-compiler no"
C = 4.06.0
D = 4.06.0 "-native-compiler no"
```

| **coq-build-ocaml** | A [fl-n=yes] | B [fl-n=no] | C [n=yes] | D [n=no]  |
|------------------------|-|-|-|-|
| real              | 2m49.756s    | 2m45.236s   | 0m33.169s | 0m32.667s |
| user              | 10m1.800s    | 10m10.356   | 2m59.352s | 2m59.456s |
| sys               | 0m53.052s    | 0m53.788s   | 0m37.452s | 0m35.896s |
| **coq-build-vo**    |              |             |           |           |
| real              | 2m45.629s    | 1m47.738s   | 2m30.453s | 2m0.606s  |
| user              | 10m29.820s   | 7m29.188s   | 9m37.728s | 8m19.564s |
| sys               | 2m57.500s    | 1m56.512s   | 2m46.096s | 1m57.520s |

native compiler is called with `-Oclassic`.

**Comment by mshinwell** [comment]
@chambart Can you try to finish this so we can at least consider it for 4.06.1?  (I like the idea of @xavierleroy 's suggestion as regards a flags field.)

**Comment by chambart** [comment]
I followed @xavierleroy's suggestion. It is a bit cleaner. The flag type is defined in Cmm, it might not be the most obvious place, but it is the most convenient dependency wise. I didn't push the change to the very end of the backend as the CSE flag is not relevant there anymore and that would increase the patch size. But I would be happy to if you think it is better that way.

File: asmcomp/asmgen.ml
Comment (xavierleroy) [inline asmcomp/asmgen.ml]: I would rather have this test done in module `CSEgen`, method `fundecl`.

File: testsuite/tests/asmgen/parsecmm.mly
Comment (xavierleroy) [inline testsuite/tests/asmgen/parsecmm.mly]: This `fun_fast_compile` field looks like a leftover from the earlier version of this GPR.

**Review by xavierleroy — CHANGES_REQUESTED** [review]
Looks good to me modulo the two remarks below.

**Comment by xavierleroy** [comment]
I'll be happy to merge once my two remarks above are addressed.

**Comment by chambart** [comment]
Back from holiday with some fixes. I rebased it on trunk.
If it is still considered for 4.06.1 I will make a branch for it. (only the last commit updating Changes has to be modified)

**Review by xavierleroy — APPROVED** [review]

File: Changes
Comment (xavierleroy) [inline Changes]: Can I please be listed as reviewer?

File: asmcomp/cmmgen.ml
Comment (xavierleroy) [inline asmcomp/cmmgen.ml]: It feels like you're reversing the logic.  "optimize for speed" means DO NOT reduce the code size.

File: asmcomp/linearize.ml
Comment (xavierleroy) [inline asmcomp/linearize.ml]: Here you're reversing the logic again.  "Reduce code size" means "fun_fast = false", not "fun_fast = true".

**Review by xavierleroy — CHANGES_REQUESTED** [review]
On reviewing again, I think your use of `Reduce_code_size` makes no sense: you're setting `Reduce_code_size` when you want to optimize for speed, and not setting it when you want to optimize for size.  I'm confused.

File: asmcomp/cmmgen.ml
Comment (chambart) [reply asmcomp/cmmgen.ml]: Sorry about that. Of course I reversed it. Thank you so much for noticing.

File: Changes
Comment (chambart) [reply Changes]: Sorry for forgetting that too.

**Comment by chambart** [comment]
By the way @xavierleroy, would you be want for coherence `fun_fast` to be renamed to something related to code size too ?

**Review by xavierleroy — APPROVED** [review]
Looks good to me!

**Comment by xavierleroy** [comment]
@chambart there is a conflict on Changes related to another of your GPRs, namely #1401.  I would fix it myself except that I'm not sure where #1401 was committed: 4.06? 4.06.1? trunk?  Could you please have a look?  Thanks!

**Comment by xavierleroy** [comment]
Concerning the `fun_fast` name: it's historical baggage, and I don't think it's worth renaming it if is remains a boolean flag, like today.  What could make sense for future developments is to carry a list of compilation options all the way to the asm emitters, like this PR does for C-- functions.

**Comment by chambart** [comment]
#1401 was merged in trunk too. It is an improvement independently of that one.

I rebased for trunk. I'm waiting for CI to approve before merging.

I'm pushing the flag propagation to assembly on my todo list.

**Comment by mshinwell** [comment]
CI has now passed and @xavierleroy has approved, so I'm merging this.

**Pull request merged**
